ArduSub/mode_emergency.cpp

```diff
@@ -0,0 +1,55 @@
+#include "Sub.h"
+
+
+bool ModeEmergency::init(bool ignore_checks)
+{
+    has_surfaced = false;
+    // initialize vertical speeds and acceleration
+    position_control->set_max_speed_accel_U_cm(-sub.get_pilot_speed_dn(), g.pilot_speed_up, g.pilot_accel_z);
+    position_control->set_correction_speed_accel_U_cmss(-sub.get_pilot_speed_dn(), g.pilot_speed_up, g.pilot_accel_z);
+
+    // initialise position and desired velocity
+    position_control->init_U_controller();
+    GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL,"Emergencey surface!");
+    return true;
+    
+    has_surfaced = false;
+
+}
+
+void ModeEmergency::run()
+{
+    // if not armed set throttle to zero and exit immediately
+    if (!motors.armed()) {
+        motors.output_min();
+        motors.set_desired_spool_state(AP_Motors::DesiredSpoolState::GROUND_IDLE);
+        attitude_control->set_throttle_out(.5,true,g.throttle_filt);
+        attitude_control->relax_attitude_controllers();
+        position_control->init_U_controller();
+        return;
+    }
+
+    // Already at surface, hold depth at surface
+    if (sub.ap.at_surface || has_surfaced) {
+        if (!has_surfaced) {
+            GCS_SEND_TEXT(MAV_SEVERITY_WARNING,"At Surface");
+            has_surfaced = true;
```

Comment:
This split-state thing is odd.  Perhaps we need a "was_at_surface" thing instead of `has_surfaced`?